text/1028-cli-hooks.rst

```diff
@@ -56,70 +69,230 @@ are always executed in WSL.
+for watching file system changes and triggering scripts.
+
+Each ``[watch.<name>]`` sub-table can have any arbitrary ``<name>`` and
+represents a single watch trigger. Multiple watch triggers can be specified as
+long as they have their own unique ``<name>``.
+
+Each watch trigger should have the following structure:
+
+* ``glob = "<path-string>"`` - specify file(s) being watched; required
+* ``script = "<command>"`` - command to be executed by the shell; required
+
+We may add more configuration options in the future to the watch triggers to
+expand functionality.
+
+The ``gel watch`` command (without needing further options) would then be used
+to start the watch process and monitor whatever is specified in the
+``gel.toml``. By default ``gel watch`` will only watch the files specified in
+the ``gel.toml`` config and execute the trigger scripts.
+
+Running ``gel watch --migrate`` will additionally monitor schema changes and
+perform real-time migrations. The ``--migrate`` option also will automatically
+retry applying migrations if it encounters network errors.
+
+This is a backwards incompatible change compared to how ``edgedb watch`` operates
+now.
+
+The output of the scripts will then appear in the same TTY as the ``gel
+watch`` command.
+
+
+Trigger Execution
+-----------------
+
+Due to the nature of monitoring changes, we cannot guarantee any particular
+order in which the watch scripts will be executed when multiple changes occur
+at once. This is independent of the source of multiple triggers: whether it is
+due to multiple files being updated or multiple watch rules matching the same
+file. The scripts are effectively triggered asynchronously and the exact order
+is an implementation detail that should not be relied upon.
+
+It is possible to trigger the same watch target multiple times in a row either
+by multiple consecutive update events to the same file or by virtue of the
+watch trigger monitoring several files which have been modified. We want to
+reduce unnecessary repeated execution of the same trigger script by debouncing
+multiple file system change events into a single trigger event.
+
+To achieve this ``gel watch`` should keep an event queue for triggered
+scripts. Any time a new trigger change is detected, the event queue is
+searched for the same trigger event. If it is found, the stale event is
+removed from the queue. The *new* event is added to the end of the queue with
+the current timestamp. This means that the events in the queue appear in
+chronological order. The head of the queue is checked periodically by ``gel
+watch`` and if the timestamp of the trigger event is more than 100ms old, the
+event is *popped* from the queue and the corresponding script is executed.
+This guarantees that no more than one copy of any trigger event is ever
+present in the queue at the same time. It also guarantees that eventually all
+triggered events will execute as long as updates triggering them actually
+stop.
```

Comment:
This might be too specific: it is describing implementation, not the behavior. And currently, it is not implemented with a single event queue.

Reply:
OK, I'll rewrite it as more abstract behavior. Just need to emphasise that the goal is to never have more than one copy of a script waiting in "debouncing" mode.